File: scripts/check-docs.ts
import { readdir, readFile, access } from 'node:fs/promises';
import { join, dirname, resolve } from 'node:path';
import { publicDirectories, publicFiles } from './build.ts';
const files:string[]=[];
async function collect(directory:string){for(const entry of await readdir(directory,{withFileTypes:true})){if(entry.name.startsWith('.')||entry.name==='private')continue;const path=join(directory,entry.name);if(entry.isDirectory())await collect(path);else if(entry.name.endsWith('.md'))files.push(path);}}
for(const directory of publicDirectories)await collect(directory);
files.push(...publicFiles.filter(f=>f.endsWith('.md')));
const errors:string[]=[];
for(const file of files){
 const text=await readFile(file,'utf8');
 if(text.split('\n').filter(line=>line.startsWith('```')).length%2)errors.push(`${file}: unclosed code fence`);
 for(const [,target] of text.matchAll(/\[[^\]]*\]\(([^)]+)\)/g)){
  if(/^[a-zA-Z][a-zA-Z0-9+.-]*:/.test(target)||target.startsWith('#'))continue;
  const local=decodeURIComponent(target.split('#')[0].split('?')[0]).replace(/^<|>$/g,'');
  if(!local)continue;
  try{await access(resolve(dirname(file),local))}catch{errors.push(`${file}: missing ${target}`)}
 }
}
if(errors.length){console.error(errors.join('\n'));process.exitCode=1}else console.log(`Checked ${files.length} public Markdown documents.`);

File: scripts/check_docs.py
#!/usr/bin/env python3
"""Check local Markdown links and code fences without network access."""
import re
from pathlib import Path
from urllib.parse import unquote
ROOT = Path(__file__).resolve().parents[1]
errors = []
files = [p for p in ROOT.rglob('*.md') if not any(x.startswith('.') or x in {'node_modules', 'private', 'dist'} for x in p.relative_to(ROOT).parts)]
for path in files:
    content = path.read_text()
    if sum(line.startswith('```') for line in content.splitlines()) % 2:
        errors.append(f'{path.relative_to(ROOT)}: unclosed code fence')
    for target in re.findall(r'\[[^\]]*\]\(([^)]+)\)', content):
        if re.match(r'^[a-zA-Z][a-zA-Z0-9+.-]*:', target) or target.startswith('#'):
            continue
        local = unquote(target.split('#')[0].split('?')[0]).strip('<>')
        if local and not (path.parent / local).exists():
            errors.append(f'{path.relative_to(ROOT)}: missing {target}')
if errors:
    print('\n'.join(errors))
    raise SystemExit(1)
print(f'Checked local links and code fences in {len(files)} Markdown files.')
